File: medium/2904-shortest-and-lexicographically-smallest-beautiful-string/solution.java
import java.util.ArrayList;
import java.util.List;

/**
 * 2904. Shortest and Lexicographically Smallest Beautiful String
 * Time: O(n^2)
 * Space: O(n)
 */
class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        // A beautiful substring can always be trimmed to start and end on a '1',
        // so only windows of k consecutive ones matter.
        List<Integer> ones = new ArrayList<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '1') {
                ones.add(i);
            }
        }
        if (ones.size() < k) {
            return "";
        }

        String best = "";
        for (int i = 0; i + k - 1 < ones.size(); i++) {
            int start = ones.get(i), end = ones.get(i + k - 1);
            String candidate = s.substring(start, end + 1);
            if (best.isEmpty()
                    || candidate.length() < best.length()
                    || (candidate.length() == best.length() && candidate.compareTo(best) < 0)) {
                best = candidate;
            }
        }
        return best;
    }
}

File: medium/2904-shortest-and-lexicographically-smallest-beautiful-string/solution.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        """
        2904. Shortest and Lexicographically Smallest Beautiful String
        Time: O(n^2)
        Space: O(n)
        """
        # A beautiful substring can always be trimmed to start and end on a '1',
        # so only windows of k consecutive ones matter.
        ones = [i for i, ch in enumerate(s) if ch == "1"]
        if len(ones) < k:
            return ""

        best = ""
        for i in range(len(ones) - k + 1):
            start, end = ones[i], ones[i + k - 1]
            candidate = s[start:end + 1]
            if not best or len(candidate) < len(best) or (
                len(candidate) == len(best) and candidate < best
            ):
                best = candidate
        return best
